src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs

```diff
@@ -6425,5 +7198,251 @@ public static class Extensions
+    {
+        await foreach (var i in new C())
+        {
+            Console.Write(i);
+        }
+    }
+
+    private static C.Enumerator GetAsyncEnumerator(this C self) => new C.Enumerator();
+}
+
+public class C
+{
+    public sealed class Enumerator
+    {
+        public int Current { get; private set; }
+        public Task<bool> MoveNextAsync() => Task.FromResult(Current++ != 3);
+    }
+}";
+            var comp = CreateCompilationWithMscorlib46(source, options: TestOptions.DebugExe, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "123");
+        }
+
+        [Fact]
+        public void TestGetAsyncEnumeratorPatternViaAccessiblePrivateExtensionInNestedClass()
+        {
+            var source = @"
+using System;
+using System.Threading.Tasks;
+public static class Program
+{
+    public static class Inner
+    {
+        public static async Task Main()
+        {
+            await foreach (var i in new C())
+            {
+                Console.Write(i);
+            }
+        }
+    }
+
+    private static C.Enumerator GetAsyncEnumerator(this C self) => new C.Enumerator();
+}
+
+public class C
+{
+    public sealed class Enumerator
+    {
+        public int Current { get; private set; }
+        public Task<bool> MoveNextAsync() => Task.FromResult(Current++ != 3);
+    }
+}";
+            var comp = CreateCompilationWithMscorlib46(source, options: TestOptions.DebugExe, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "123");
+        }
+
+        [Fact]
+        public void TestGetAsyncEnumeratorPatternViaInaccessiblePrivateExtension()
+        {
+            var source = @"
+using System;
+using System.Threading.Tasks;
+public class C
+{
+    public static async Task Main()
+    {
+        await foreach (var i in new C())
+        {
+            Console.Write(i);
+        }
+    }
+    public sealed class Enumerator
+    {
+        public int Current { get; private set; }
+        public Task<bool> MoveNextAsync() => Task.FromResult(Current++ != 3);
+    }
+}
+public static class Extensions
+{
+    private static C.Enumerator GetAsyncEnumerator(this C self) => new C.Enumerator();
+}
+";
+            CreateCompilation(source, parseOptions: TestOptions.RegularPreview)
+                .VerifyDiagnostics(
+                    // (8,33): error CS8411: Asynchronous foreach statement cannot operate on variables of type 'C' because 'C' does not contain a suitable public instance or extension definition for 'GetAsyncEnumerator'
+                    //         await foreach (var i in new C())
+                    Diagnostic(ErrorCode.ERR_AwaitForEachMissingMember, "new C()").WithArguments("C", "GetAsyncEnumerator").WithLocation(8, 33)
+                    );
+        }
+
+        [Fact]
+        public void TestGetAsyncEnumeratorPatternViaExtensionWithRefReturn()
```

Comment:
I was expecting this to error, as we would for a `ref`-returning `GetAsyncEnumerator` today.

Reply:
Hmm, my mistake. My test case was incorrect, we do allow this.

---
In reply to: [443873802](https://github.com/dotnet/roslyn/pull/45160#discussion_r443873802) [](ancestors = 443873802)